File: Examples/VisualBasic/Exchange/FilterTasksFromServer.vb
Imports System.IO
Imports Aspose.Email.Mail
Imports Aspose.Email.Outlook
Imports Aspose.Email.Pop3
Imports Aspose.Email
Imports Aspose.Email.Mime
Imports Aspose.Email.Imap
Imports System.Configuration
Imports System.Data
Imports Aspose.Email.Mail.Bounce
Imports Aspose.Email.Exchange

Namespace Aspose.Email.Examples.VisualBasic.Email.Exchange
    Public Class FilterTasksFromServer
        Public Shared Sub Run()
            ' ExStart:FilterTasksFromServer
            ' Set mailboxURI, Username, password, domain information
            Dim mailboxUri As String = "https://ex2010/ews/exchange.asmx"
            Dim username As String = "test.exchange"
            Dim password As String = "pwd"
            Dim domain As String = "ex2010.local"
            Dim credentials As New NetworkCredential(username, password, domain)
            Dim client As IEWSClient = EWSClient.GetEWSClient(mailboxUri, credentials)

            Dim queryBuilder As ExchangeQueryBuilder = Nothing
            Dim query As MailQuery = Nothing
            Dim fetchedTask As ExchangeTask = Nothing
            Dim messageInfoCol As ExchangeMessageInfoCollection = Nothing
            client.TimezoneId = "Central Europe Standard Time"
            Dim values As Array = [Enum].GetValues(GetType(ExchangeTaskStatus))

            'Now retrieve the tasks with specific statuses
            For Each status As ExchangeTaskStatus In values
                queryBuilder = New ExchangeQueryBuilder()
                queryBuilder.TaskStatus.Equals(status)
                query = queryBuilder.GetQuery()
                messageInfoCol = client.ListMessages(client.MailboxInfo.TasksUri, query)
                fetchedTask = client.FetchTask(messageInfoCol(0).UniqueUri)
            Next

            'retrieve all other than specified
            For Each status As ExchangeTaskStatus In values
                queryBuilder = New ExchangeQueryBuilder()
                queryBuilder.TaskStatus.NotEquals(status)
                query = queryBuilder.GetQuery()
                messageInfoCol = client.ListMessages(client.MailboxInfo.TasksUri, query)
            Next

            'specifying multiple criterion
            Dim selectedStatuses As ExchangeTaskStatus() = New ExchangeTaskStatus() {ExchangeTaskStatus.Completed, ExchangeTaskStatus.InProgress}
            queryBuilder = New ExchangeQueryBuilder()
            queryBuilder.TaskStatus.[In](selectedStatuses)
            query = queryBuilder.GetQuery()
            messageInfoCol = client.ListMessages(client.MailboxInfo.TasksUri, query)

            queryBuilder = New ExchangeQueryBuilder()
            queryBuilder.TaskStatus.NotIn(selectedStatuses)
            query = queryBuilder.GetQuery()
            messageInfoCol = client.ListMessages(client.MailboxInfo.TasksUri, query)
            'ExEnd:FilterTasksFromServer
        End Sub
    End Class
End Namespace

File: Examples/Aspose.Email.Examples/EWS/FilterTasksFromServer.cs
using Aspose.Email.Clients.Exchange;
using Aspose.Email.Clients.Exchange.WebService;
using Aspose.Email.Tools.Search;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Text;

namespace Aspose.Email.Examples.EWS
{
    class FilterTasksFromServer
    {
        public static void Run()
        {
            // Set mailboxURI, Username, password, domain information
            string mailboxUri = "https://ex2010/ews/exchange.asmx";
            string username = "test.exchange";
            string password = "pwd";
            string domain = "ex2010.local";
            NetworkCredential credentials = new NetworkCredential(username, password, domain);
            IEWSClient client = EWSClient.GetEWSClient(mailboxUri, credentials);

            ExchangeQueryBuilder queryBuilder = null;
            MailQuery query = null;
            ExchangeTask fetchedTask = null;
            ExchangeMessageInfoCollection messageInfoCol = null;
            client.TimezoneId = "Central Europe Standard Time";
            Array values = Enum.GetValues(typeof(ExchangeTaskStatus));

            //Now retrieve the tasks with specific statuses
            foreach (ExchangeTaskStatus status in values)
            {
                queryBuilder = new ExchangeQueryBuilder();
                queryBuilder.TaskStatus.Equals(status);
                query = queryBuilder.GetQuery();
                messageInfoCol = client.ListMessages(client.MailboxInfo.TasksUri, query);
                fetchedTask = client.FetchTask(messageInfoCol[0].UniqueUri);
            }

            //retrieve all other than specified
            foreach (ExchangeTaskStatus status in values)
            {
                queryBuilder = new ExchangeQueryBuilder();
                queryBuilder.TaskStatus.NotEquals(status);
                query = queryBuilder.GetQuery();
                messageInfoCol = client.ListMessages(client.MailboxInfo.TasksUri, query);
            }

            //specifying multiple criterion
            ExchangeTaskStatus[] selectedStatuses = new ExchangeTaskStatus[]
                    {
            ExchangeTaskStatus.Completed,
            ExchangeTaskStatus.InProgress
                    };
            queryBuilder = new ExchangeQueryBuilder();
            queryBuilder.TaskStatus.In(selectedStatuses);
            query = queryBuilder.GetQuery();
            messageInfoCol = client.ListMessages(client.MailboxInfo.TasksUri, query);

            queryBuilder = new ExchangeQueryBuilder();
            queryBuilder.TaskStatus.NotIn(selectedStatuses);
            query = queryBuilder.GetQuery();
            messageInfoCol = client.ListMessages(client.MailboxInfo.TasksUri, query);
        }
    }
}
